src/c4/ofpstream.hh

```diff
@@ -0,0 +1,86 @@
+//----------------------------------*-C++-*----------------------------------//
+/*!
+ * \file   c4/ofpstream.hh
+ * \author Mike Buksas
+ * \brief  Define class ofpstream
+ * \note   Copyright (C) 2016-2018 Los Alamos National Security, LLC.
+ *         All rights reserved.
+ */
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+
+#ifndef c4_ofpstream_hh
+#define c4_ofpstream_hh
+
+#include <fstream>
+#include <vector>
+
+namespace rtt_c4 {
+
+//===========================================================================//
+/*!
+ * \class ofpstream
+ *
+ * Output stream for serializing output from all MPI ranks.
+ *
+ * This class is an output stream that serialized output from all MPI ranks.
+ * Output is inserted to the stream as with any other output stream. When the
+ * send() function is invoked, which *must be done synchronously across all
+ * processors*, all output inserted to the stream since the last call to send()
+ * is printed to the associated file, by MPI rank. In other words, all output
+ * inserted to an instance of ofpstream by rank 0 since the last call to send()
+ * is printed to the associated file, then all output inserted to the instance of
+ * ofpstream by rank 1 is printed to cout, and so on for all processors. There
+ * is no requirement that any processor have actually inserted any output to
+ * the stream since the last cal to send(), nor does the output from any
+ * processor have to resemble that from any other processor in any way (though
+ * it very often will.)
+ *
+ * Because this requires that output be buffered in the stream for each rank
+ * between send() calls, memory consumption might become significant if
+ * very large blocks of output are generated by MPI ranks between calls to
+ * send(). In other words, this is for TL;DR type messages to a file.
+ *
+ * Memory allocated for buffering is retained between calls unless the
+ * shrink_to_fit function is called, which releases all memory not needed
+ * for the current buffered output. In typical usage, this means the stream
+ * will grow its buffer size silently until it is big enough for the largest
+ * blocks of output being generated, then remain there through the rest
+ * of the scope containing the stream, unless shrink_to_fit is called.
+ *
+ * \example c4/test/tstofpstream.cc
+ */
+//===========================================================================//
+
+class ofpstream : public std::ostream {
+public:
+  ofpstream(std::string const &filename);
+
+  void send() { sb_.send(); }
+  void shrink_to_fit() { sb_.shrink_to_fit(); }
+
+private:
+  struct mpibuf : public std::streambuf {
+
+    ~mpibuf() { send(); }
```

Comment:
Can `send()` throw? (that would be bad). Can we protect against this?

Reply:
Yes, because it allocates storage for its buffer and it does I/O, both of which are operations than can throw an exception.  I can add code to trap these, but I'd need to know how you would like these handled, if not as fatal exceptions.

Reply:
@alexrlongne Do you have any thoughts on this?  Should we just allow the thrown exception to be fatal?

Reply:
Here's some additional information about why throwing dtors are bad.  http://bin-login.name/ftp/pub/docs/programming_languages/cpp/cffective_cpp/MEC/MI11_FR.HTM

Reply:
Another possibility is to have the destructor simply discard any unflushed output. If this is done, the destructor does nothing but free its buffer, and freeing memory is not supposed to throw an exception.

Reply:
@kgbudge That sounds like a good plan.  Can you implement the nothrow dtor option that simply discards the unflushed output?

Reply:
Done.